File: Solutions/JavaScript/student.js
class Student {
    constructor(name, grades) {
        this.name = name
        this.grades = grades
    }

    get averageGrade() {
        if (this.grades.length === 0) {
            return 0
        }
        let sum = 0
        for (const grade of this.grades) {
            sum += grade
        }
        return sum / this.grades.length
    }

    get letterGrade() {
        const averageGrade = this.averageGrade
        if (averageGrade >= 90) {
            return 'A'
        } else if (averageGrade >= 80) {
            return 'B'
        } else if (averageGrade >= 70) {
            return 'C'
        } else if (averageGrade >= 60) {
            return 'D'
        } else {
            return 'F'
        }
    }
}

class StudentTracker {
    constructor() {
        this.students = []
    }

    addStudent(name, grades) {
        const student = new Student(name, grades)
        this.students.push(student)

        return student
    }

    getStudentsWithLetterGrade(letterGrade) {
        const studentsWithLetterGrade = []
        for (const student of this.students) {
            if (student.letterGrade === letterGrade) {
                studentsWithLetterGrade.push(student)
            }
        }
        return studentsWithLetterGrade
    }
}

module.exports = { Student, StudentTracker }

File: Solutions/Kotlin/src/main/kotlin/Student.kt
class Student(val name: String, val grades: List<Int>) {
    val averageGrade: Double
        get() {
            if (grades.isEmpty()) {
                return 0.0
            }
            var sum = 0.0
            for (grade in grades) {
                sum += grade
            }
            return sum / grades.size
        }

    val letterGrade: String
        get() {
            val averageGrade = averageGrade
            return when {
                averageGrade >= 90 -> "A"
                averageGrade >= 80 -> "B"
                averageGrade >= 70 -> "C"
                averageGrade >= 60 -> "D"
                else -> "F"
            }
        }
}

class StudentTracker {
    private val students = mutableListOf<Student>()

    val studentCount: Int
        get() = students.size

    fun addStudent(name: String, grades: List<Int>): Student {
        val student = Student(name, grades)
        students.add(student)
        return student
    }

    fun getStudentsWithLetterGrade(letterGrade: String): List<Student> {
        val studentsWithLetterGrade = mutableListOf<Student>()
        for (student in students) {
            if (student.letterGrade == letterGrade) {
                studentsWithLetterGrade.add(student)
            }
        }
        return studentsWithLetterGrade
    }
}
